File: src/org/symbolmath/ast/ASTElement.java
package org.symbolmath.ast;

import java.util.List;

public abstract class ASTElement {
  protected ASTElement parent;

  public ASTElement() {
  }

  public ASTElement getParent() {
    return parent;
  }

  public void setParent(ASTElement parent) {
    this.parent = parent;
  }

  public abstract void replaceChild(ASTElement element, ASTElement newElement);

  public abstract String toString();

  public abstract List<? extends ASTElement> getChildren();
}

File: src/main/java/org/symbolmath/ast/ASTElement.kt
package org.symbolmath.ast

abstract class ASTElement {
    var parent: ASTElement? = null
    abstract val children: List<ASTElement>
    abstract fun replaceChild(element: ASTElement, newElement: ASTElement)
    abstract fun negate(): ASTElement
    open fun isZero() = false
    abstract override fun toString(): String
    abstract override fun equals(other: Any?): Boolean
}